test/test_mbarrier.cpp

```diff
@@ -0,0 +1,146 @@
+    mbarrier->setMemoryType(MemoryType::Shared);
+    kir::Allocate* mbarrier_alloc =
+        IrBuilder::create<kir::Allocate>(mbarrier, MemoryType::Shared);
+    dynamic_smem_allocations.push_back(mbarrier_alloc);
+
+    Val* mbarrier_address = SimplifyingIrBuilder::mulExpr(
+        dynamic_smem_allocations.at(0)->size(),
+        dataTypeSize(dynamic_smem_allocations.at(0)->buffer()->dtype()));
+    mbarrier_alloc->setAddress(mbarrier_address);
+
+    auto smem_alloc_it = std::find_if(
+        top_level_exprs.begin(), top_level_exprs.end(), [](Expr* expr) {
+          if (auto alloc = dynamic_cast<kir::Allocate*>(expr)) {
+            if (auto tv = dynamic_cast<TensorView*>(alloc->buffer())) {
+              return tv->getMemoryType() == MemoryType::Shared;
+            } else {
+              return false;
+            }
+          }
+          return false;
+        });
+    smem_alloc_it++;
+    ASSERT_NE(smem_alloc_it, top_level_exprs.end());
+    smem_alloc_it = top_level_exprs.insert(smem_alloc_it, mbarrier_alloc);
+
+    // Indexing mbarrier
+    auto mbarrier_smem_addr = IrBuilder::create<Val>(DataType::SMemAddress);
+    IrBuilder::create<UnaryOp>(
+        UnaryOpType::ToUnsignedSmemAddr,
+        mbarrier_smem_addr,
+        IrBuilder::metadataExpr(mbarrier));
+    auto mbarrier_index =
+        IrBuilder::create<kir::TensorIndex>(mbarrier, mbarrier_smem_addr);
+
+    // Initialize mbarrier
+    smem_alloc_it++;
+    ASSERT_NE(smem_alloc_it, top_level_exprs.end());
+    auto init = IrBuilder::create<kir::MBarrierInit>(
+        mbarrier_index, IrBuilder::create<Val>(1024, DataType::UInt32));
+    top_level_exprs.insert(smem_alloc_it, init);
+
+    // Arrive and wait
+    auto sync_it = std::find_if(
+        top_level_exprs.begin(), top_level_exprs.end(), [](Expr* expr) {
+          return expr->isA<kir::BlockSync>();
+        });
+    ASSERT_NE(sync_it, top_level_exprs.end());
+    auto state = IrBuilder::create<Val>(DataType::UInt);
+    auto alloc_state = IrBuilder::create<kir::Allocate>(
+        state, MemoryType::Local, kernel->oneVal());
+    auto arrive = IrBuilder::create<kir::MBarrierArrive>(state, mbarrier_index);
+    auto wait = IrBuilder::create<kir::MBarrierWait>(mbarrier_index, state);
+    *sync_it = wait;
+    sync_it = top_level_exprs.insert(sync_it, arrive);
+    top_level_exprs.insert(sync_it, alloc_state);
+
+    // Invalidate mbarrier
+    auto invalidate =
+        IrBuilder::create<kir::MBarrierInvalidate>(mbarrier_index);
+    top_level_exprs.push_back(invalidate);
+  });
+
+  fe.compileFusion(&fusion);
+
+  std::unordered_set<const std::type_info*> remaining_mbarrier_exprs{
+      &typeid(kir::MBarrierInit),
+      &typeid(kir::MBarrierArrive),
+      &typeid(kir::MBarrierWait),
+      &typeid(kir::MBarrierInvalidate)};
+  for (auto expr : fe.kernel()->topLevelExprs()) {
+    remaining_mbarrier_exprs.erase(&typeid(*expr));
+  }
+  EXPECT_TRUE(remaining_mbarrier_exprs.empty());
```

Comment:
what's the purpose of this part?  Does it ensure that all `MBarrier` expressions are correctly integrated into the `kir`? I saw other test cases are directly checking kernel string, e.g. `FusionCodegenAllocatedScalars_CUDA`

Reply:
Yes, that's it. And directly checking kernel string is also a way.